core/ia32.asm

```diff
@@ -0,0 +1,351 @@
+    mov reg_ret_16, %1
+    ret
+%endmacro
+%macro function_set_segment 1
+    function set_kernel_%+%1
+    mov %1, reg_arg1_16
+    ret
+%endmacro
+
+section .text
+
+struc qword_struct
+    .lo      resd 1
+    .hi      resd 1
+endstruc
+
+struc vcpu_state
+    ._rax    resq 1
+    ._rcx    resq 1
+    ._rdx    resq 1
+    ._rbx    resq 1
+    ._rsp    resq 1
+    ._rbp    resq 1
+    ._rsi    resq 1
+    ._rdi    resq 1
+    ._r8     resq 1
+    ._r9     resq 1
+    ._r10    resq 1
+    ._r11    resq 1
+    ._r12    resq 1
+    ._r13    resq 1
+    ._r14    resq 1
+    ._r15    resq 1
+endstruc
+
+struc cpuid_args
+    ._eax    resd 1
+    ._ecx    resd 1
+    ._edx    resd 1
+    ._ebx    resd 1
+endstruc
+
+function __nmi
+    int 2h
+    ret
+
+function __fls
+    xor reg_ret_32, reg_ret_32
+    bsr reg_ret_32, reg_arg1_32
+    ret
+
+function asm_cpuid
+%ifidn __BITS__, 64
+    push rbx
+    mov r8, reg_arg1
+    mov rax, [r8 + vcpu_state._rax]
+    mov rcx, [r8 + vcpu_state._rcx]
+    cpuid
+    mov [r8 + vcpu_state._rax], rax
+    mov [r8 + vcpu_state._rbx], rbx
+    mov [r8 + vcpu_state._rcx], rcx
+    mov [r8 + vcpu_state._rdx], rdx
+    pop rbx
+    ret
+%else
+    push ebx
+    push esi
+    mov esi, reg_arg1
+    mov eax, [esi + vcpu_state._rax]
+    mov ecx, [esi + vcpu_state._rcx]
+    cpuid
+    mov [esi + vcpu_state._rax], eax
+    mov [esi + vcpu_state._rbx], ebx
+    mov [esi + vcpu_state._rcx], ecx
+    mov [esi + vcpu_state._rdx], edx
+    pop esi
+    pop ebx
+    ret
+%endif
+
+function asm_btr
+    lock btr [reg_arg1], reg_arg2
+    ret
+
+function asm_bts
+    lock bts [reg_arg1], reg_arg2
+    ret
+
+function asm_disable_irq
+    cli
+    ret
+
+function asm_enable_irq
+    sti
+    ret
+
+function asm_fxinit
+    finit
+    ret
+
+function asm_fxrstor
+    fxrstor [reg_arg1]
+    ret
+
+function asm_fxsave
+    fxsave [reg_arg1]
+    ret
+
+function asm_rdmsr
+%ifidn __BITS__, 64
+    mov rcx, reg_arg1
+    rdmsr
+    shl rdx, 32
+    or reg_ret, rdx
+    ret
+%else
+    mov ecx, reg_arg1
+    rdmsr
+    mov [reg_arg2 + qword_struct.lo], eax
+    mov [reg_arg2 + qword_struct.hi], edx
+    ret
+%endif
+
+function asm_rdtsc
+%ifidn __BITS__, 64
+    rdtsc
+    shl rdx, 32
+    or reg_ret, rdx
+    ret
+%else
+    rdtsc
+    mov [reg_arg1 + qword_struct.lo], eax
+    mov [reg_arg1 + qword_struct.hi], edx
+    ret
+%endif
+
+function asm_wrmsr
+%ifidn __BITS__, 64
+    push rbx
+    mov rbx, reg_arg2
+    mov rcx, reg_arg1
+    mov eax, ebx
+    mov rdx, rbx
+    shr rdx, 32
+    wrmsr
+    pop rbx
+    ret
+%else
+    push edi
+    push esi
+    mov edi, [reg_arg2 + qword_struct.lo]
+    mov esi, [reg_arg2 + qword_struct.hi]
+    mov ecx, reg_arg1
+    mov eax, edi
+    mov edx, esi
+    wrmsr
+    pop esi
+    pop edi
+    ret
+%endif
+
+function get_kernel_tr_selector
+    str reg_ret_16
+    ret
+
+function get_kernel_ldt
+    sldt reg_ret_16
+    ret
+
+function get_kernel_gdt
+    sgdt [reg_arg1]
```

Comment:
nit: Does `sgdt` not require the operand to be labeled `fword ptr` like `lgdt` does?

Reply:
Interesting. But it makes sense, because `lgdt`, etc. don't really use other (standard) operand sizes, so removing `fword ptr` doesn't really create any ambiguity.